openedx/features/content_type_gating/models.py

```diff
@@ -57,36 +50,24 @@ class ContentTypeGatingConfig(StackedConfigurationModel):
     )
 
     @classmethod
-    def has_full_access_role_in_masquerade(cls, user, course_key, course_masquerade, student_masquerade,
-                                           user_partition):
+    def has_full_access_role_in_masquerade(cls, user, course_key, course_masquerade):
         """
         The roles of the masquerade user are used to determine whether the content gate displays.
         The gate will not appear if the masquerade user has any of the following roles:
         Staff, Instructor, Beta Tester, Forum Community TA, Forum Group Moderator, Forum Moderator, Forum Administrator
         """
-        if student_masquerade:
+        if is_masquerading_as_specific_student(user, course_key):
             # If a request is masquerading as a specific user, the user variable will represent the correct user.
             if user and user.id and has_staff_roles(user, course_key):
                 return True
-        elif user_partition:
-            # If the current user is masquerading as a generic student in a specific group,
-            # then return the value based on that group.
-            masquerade_group = get_masquerading_user_group(course_key, user, user_partition)
-            if masquerade_group is None:
-                audit_mode_id = settings.COURSE_ENROLLMENT_MODES.get(CourseMode.AUDIT, {}).get('id')
-                # We are checking the user partition id here because currently content
-                # cannot have both the enrollment track partition and content gating partition
-                # configured simultaneously. We may change this in the future and allow
-                # configuring both partitions on content and selecting both partitions in masquerade.
-                if course_masquerade.user_partition_id == ENROLLMENT_TRACK_PARTITION_ID:
-                    return course_masquerade.group_id != audit_mode_id
-            elif masquerade_group is FULL_ACCESS:
-                return True
-            elif masquerade_group is LIMITED_ACCESS:
-                return False
+        elif (is_masquerading_as_staff(user, course_key) or
+              is_masquerading_as_full_access(user, course_key, course_masquerade) or
+              is_masquerading_as_non_audit_enrollment(user, course_key, course_masquerade)):
+            return True
+        return False
```

Comment:
I'm fine punting on this for now, as I think this is a "pre-existing condition", but figured I'd document it since I looked into it...

This _ever_ _so_ _slightly_ changes the method signature here w/ the possible return values...

The previous logic was:
- `return True`: if we can positively confirm that the user is masquerading as a (class of) user who has full access
- `return False`: if we can positively confirm that the user is _not_ masquerading as a (class of) user who has full access
- `return None`: (implicitly, w/ no `return` value) if we can _not_ positively confirm whether or not the user is masquerading as a (class of) user who has full access

Now, the two `false-y`  cases are combined and we:
- `return False`: if we can _not_ positively confirm that the user is masquerading as a (class of) user who has full access

Immediately, this seems trivial, but the side effect in the calling function is that even if the user is explicitly masquerading as a user with limited access (and we should enable the content gating here), we still go on to perform superfluous checks that _could_ result in the lock _not_ being shown after all.


To work around this, and allow ourselves to bail earlier in the `enabled_for_enrollment` call, we can refactor this to:

```python
# pardon my brevity...
if not course_masquerade:
    return False
if specific_student and staff:
    return True
if staff or full_access or non_audit:
    return True
if limited_access:
    return False
return None
```

and then when we call it:
```py
has_full_access = cls.has_full_access_role_in_masquerade(user, course_key, course_masquerade)
if has_full_access is not None:
    return not has_full_access
# else, fall through

But again, fine to defer.

Reply:
I think you're right... The only difference between early exiting if the user is masquerading as a generic limited access learner is that you miss 3 further checks: "is the user in a holdback", "is the user in audit mode", and "was FBE enabled at the right datetime".

The first and third are effectively skipped if we are masquerading as a generic user. So only the second check would be missed if we early exit.

And that would actually probably be a bugfix, since if the staff user happens to have a verified enrollment, we don't want to know about it, since they asked to masquerade as a limited access person.

Let me think if I want to fix that or just let the sleeping dog lie.

Though, in your code snippet above, you return False if we aren't masquerading and you probably want None, right?

Reply:
> Though, in your code snippet above, you return False if we aren't masquerading and you probably want None, right?

Yes, good catch.

Reply:
> And that would actually probably be a bugfix, since if the staff user happens to have a verified enrollment, we don't want to know about it, since they asked to masquerade as a limited access person.

@mikix Actually, looking at this again, while this is an existing bug _here_, this PR would introduce a regression on the courseware_api and subsequently the masquerade feature in the courseware MFE because it explicitly _does_ check for that case (that's part of the reason I built out all those new helpers recently) and ensures it bails early if you're explicitly masquerading as a `LIMITED_ACCESS` user.

Reply:
OK. I refactored quite a bit again. Now:
- has_full_access_role_in_masquerade lives in the helpers.py file and returns a tri-state (None, True, or False) based on masquerade state (and it checks if masquerading at all).
- I moved the bulk of the bodies of the two gating `enabled_for_enrollment` calls into a new `enrollment_date_for_fbe` helper, which is also hopefully more readable than previous versions.
- By using this new helper in the CourseDurationLimitConfig class, I believe I also fixed a bug where it wasn't checking for `correct_modes_for_fbe` in this code path.